light-clients/icsxx-cf-solana/src/header.rs

```diff
@@ -1,68 +1,146 @@
+			return Err(Error::LastShredNotLastInSlot)
+		}
 
-		// TODO: replace with a return error
-		assert!(last_data_shred.last_in_slot());
-		let _parent_slot = last_data_shred.parent().unwrap();
 		let completed_data_indexes = data_shreds
 			.iter()
 			.filter_map(|s| if s.data_complete() { Some(s.index()) } else { None })
 			.collect::<BTreeSet<_>>();
 		let consumed = last_data_shred.index() + 1;
 		let completed_ranges = get_completed_data_ranges(0, &completed_data_indexes, consumed);
 		let slot = self.slot();
-		let entries = get_slot_entries_in_block(slot, completed_ranges, &data_shreds).unwrap();
-		let slot_entries = entries;
-
-		let blockhash = slot_entries
-			.last()
-			.map(|entry| entry.hash)
-			.unwrap_or_else(|| panic!("Rooted slot {slot:?} must have blockhash"));
-		blockhash
+		let entries = get_slot_entries_in_block(slot, completed_ranges, &data_shreds)?;
+		let blockhash = entries.last().map(|entry| entry.hash).ok_or(Error::EntriesAreEmpty)?;
+
+		Ok(blockhash)
+	}
+}
+
+/// An immutable array of shreds that have the following properties:
+/// 1. Is non-empty
+/// 2. Is sorted
+/// 3. Doesn't contain duplicates
+/// 4. All shreds have the same slot
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct PreCheckedShreds(Vec<Shred>);
+
+impl PreCheckedShreds {
+	/// Returns the (common) shreds' slot. Since the shreds are pre-checked, the array contains at
+	/// least one shred and all the shreds have same slot.
+	pub(crate) fn slot(&self) -> Slot {
+		self.0[0].slot()
 	}
 }
 
-impl<PK: PubKey> From<Header<PK>> for proto::Header {
-	fn from(value: Header<PK>) -> Self {
+impl Deref for PreCheckedShreds {
+	type Target = [Shred];
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+impl IntoIterator for PreCheckedShreds {
+	type Item = Shred;
+	type IntoIter = alloc::vec::IntoIter<Shred>;
+
+	fn into_iter(self) -> Self::IntoIter {
+		self.0.into_iter()
+	}
+}
+
+impl TryFrom<Vec<Shred>> for PreCheckedShreds {
+	type Error = Error;
+
+	/// Creates a new `PreCheckedShreds` from a `Vec<Shred>`. The input shreds must satisfy the
+	/// properties of the `PreCheckedShreds` type (see the `PreCheckedShreds`'s documentation).
+	fn try_from(shreds: Vec<Shred>) -> Result<Self, Self::Error> {
+		if shreds.is_empty() {
+			return Err(Error::ShardsAreEmpty)
+		}
+
+		let mut shreds_set = BTreeSet::new();
+		let mut prev_index = None;
+		let slot = shreds[0].slot();
+
+		for shred in &shreds {
+			let index = shred.index();
+
+			// Ensure the shreds are from the same slot.
+			if shred.slot() != slot {
+				return Err(Error::ShredsFromDifferentSlots);
+			}
+
+			// Ensure the shreds are sorted.
+			if let Some(prev_index) = prev_index {
+				if prev_index >= index {
+					return Err(Error::ShredsNotSorted);
+				}
+			}
+
+			// Ensure the shreds don't contain duplicates.
+			if !shreds_set.insert(index) {
+				return Err(Error::ShredsContainDuplicates);
+			}
+
+			prev_index = Some(index);
+		}
+
+		Ok(Self(shreds))
```

Comment:
Another thing is that we seem to only care about data shreds so maybe we should filter non-data shreds out?

```suggestion
		shreds.retain(Shred::is_data);
		Ok(Self(shreds))
```

Reply:
Added a return error if some shreds are not data